server/internal/worker/worker.go

```diff
@@ -0,0 +1,136 @@
+package worker
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// A Job represents a unit of work to be done by a worker in the pool.
+type Job struct {
+	// job == nil is used as a sentinal value to instruct workers to quit
+	job    interface{}
+	commit func(result interface{})
+}
+
+// NewJob creates a new Job. job will be provided as the input to a worker's Do
+// method, and once the Do method returns, its return value will be passed as
+// the argument to commit.
+func NewJob(job interface{}, commit func(result interface{})) Job {
+	if job == nil {
+		panic("nil job")
+	}
+	return Job{job: job, commit: commit}
+}
+
+// An Idler performs a unit of background work when Idle is called.
+type Idler interface {
+	// Idle performs a unit of background work.
+	Idle()
+}
+
+// A Worker is capable of executing jobs.
+type Worker interface {
+	// Do performs the given job and returns the result.
+	Do(job interface{}) (result interface{})
+}
+
+// A Pool is a handle on a pool of worker goroutines that can execute jobs or
+// idle work.
+type Pool struct {
+	jobs    chan Job
+	workers int // number of workers (not including idlers)
+	wg      sync.WaitGroup
+}
+
+// NewPool constructs a new Pool from the given workers. Each worker is run in
+// its own goroutine. The workers wait jobs to be submitted and execute those
+// jobs as they come in.
+func NewPool(workers ...Worker) *Pool {
```

Comment:
Using varargs is more idiomatic, and it allows users to call it like `NewPool(a, b, c)` if they want.